### PR TITLE
Useful Bang

### PR DESCRIPTION
**Pull request by ameyjadiye**
This file tells about the Bang(!) usage, very usefull for the linux admins and programmers who always needs to be on terminals.

### CODE REVIEW

**Comment by jucrouzet** [comment]
IMHO, that's something very common and every shell user, even beginner, already used it.
:-1:

**Comment by dborzov** [comment]
Was helpful to me, I like it :+1:

**Comment by ameyjadiye** [comment]
@jucrouzet glad you already used it but  [thoughtbot/til](https://github.com/thoughtbot/til)  repo description says **"We are constantly learning new things. This is a repo to share our learnings"**

When i learned  working of Bang(!), i found it awesome and shared it thinking this can be useful to someone else, thanks @dborzov for showing your interest, i'm happy you found it helpful :smiley:

**Comment by stites** [comment]
My two cents — I like it! I didn't realize you that there was more than `!!`, I guess I just found `!!` and reverse-o-matic and was set. Actually I still might be set, but still! Cool to learn!

**Comment by pirj** [comment]
First of all - it is bash and zsh specific, won't work in ksh. Second, this is evil, check 'sudo, you are doing it wrong' video.

**Comment by muescha** [comment]
just because its `evil` we can not close the eyes that the `!!` exists ;)

:+1: from me

**Comment by ananthchelladurai** [comment]
Was definitely helpful :+1: from me.

**Comment by JoelQ** [comment]
Thoughts @thoughtbot/shell ?

File: bash/sudo_bang.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Useful *Bang*
+
+I found another great thing bash console provides for the lazy programmers ;) Most of the time we need to repeat the commands we entered in linux terminal and many hardworking engineers typing it again and again ;)
+
+Linux provides many tools which remembers history and on top of that we can make our life easy.
+
+Usually i work on servers and whenever try to fire few root access commands and it dont allow me!! -_-,feels insulting. Anyway sudo !! is here to rescue you :)
```
Comment (gabebw) [inline bash/sudo_bang.md]: Can we rephrase this paragraph? Maybe: "Bang commands make it easy to re-run commands as root if you forget to use `sudo`."

File: bash/sudo_bang.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Useful *Bang*
+
+I found another great thing bash console provides for the lazy programmers ;) Most of the time we need to repeat the commands we entered in linux terminal and many hardworking engineers typing it again and again ;)
+
+Linux provides many tools which remembers history and on top of that we can make our life easy.
```
Comment (mike-burns) [inline bash/sudo_bang.md]: Linux is a kernel. Are you talking about your shell?

**Comment by mike-burns** [comment]
Drop the emoticons.

Can you link to documentation?

As mentioned elsewhere in here: this is shell-specific. Mention the shell. You put this under `bash` so I guess mention that (though I think it's from csh, and it's certainly in zsh and GNU bash too).

@pirj I don't remember `!!` being mentioned in that video, and a quick skim of the PDF doesn't mention it. Am I missing something?

**Comment by pirj** [comment]
@mike-burns it's not `!!` that's mentioned there, but rather `sudo`. You should not use sudo in cases you run services (add it to sudoers explicitly), touch files (add user tho group able to write).
So if you use user permissions in a correct way, you won't need `!!` that often.

**Comment by ameyjadiye** [comment]
Did few changes suggested by you guys, let me know if there are more modifications to be done?

**Comment by artofhuman** [comment]
:+1:

**Comment by ameyjadiye** [comment]
Hi @JoelQ / @mike-burns,
Seems nothing to change from my side, do i need to improve this article? please suggest any else its good to merge this PR :+1:

**Comment by ameyjadiye** [comment]
can someone please take this PR and merge ?

**Comment by stefannibrasil** [comment]
This repo is no longer active and it's being archived. The contents will still be available as read-only. Thank you for your contributions!

**Pull request closed without merging**